cmd/clusterimageset-updater/main.go

```diff
@@ -0,0 +1,223 @@
+		}
+		versionToPullspec[version] = pullSpec
+	}
+
+	// keep list of outdated or removed cluster image set definitions to delete
+	var toDelete []string
+	if err := filepath.WalkDir(o.outputDir, func(path string, info fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() || !strings.HasSuffix(info.Name(), "_clusterimageset.yaml") {
+			return nil
+		}
+		raw, err := ioutil.ReadFile(path)
+		if err != nil {
+			return err
+		}
+		imageset := hivev1.ClusterImageSet{}
+		if err := yaml.Unmarshal(raw, &imageset); err != nil {
+			return err
+		}
+		if imageset.Annotations != nil && imageset.Annotations["version_in"] != "" {
+			isCurrent := false
+			versionIn := imageset.Annotations["version_in"]
+			for version := range autoPools {
+				if version == versionIn {
+					if imageset.Spec.ReleaseImage == versionToPullspec[version] {
+						isCurrent = true
+						delete(autoPools, version)
+						delete(versionToPullspec, version)
+					}
+					break
+				}
+			}
+			if !isCurrent {
+				toDelete = append(toDelete, path)
+				return nil
+			}
+		}
+		return nil
+	}); err != nil {
+		logrus.WithError(err).Fatal("Failed to get list of clusterpools setting `version_in`")
+	}
+
+	// any remaining items in autopools/versionToPullspec need to be updated
+	for version, pullspec := range versionToPullspec {
+		name, err := nameFromPullspec(pullspec, version)
+		if err != nil {
+			// this shouldn't happen
+			logrus.WithError(err).Fatalf("Failed to generate clusterimageset name for version %s", version)
+		}
+		clusterimageset := hivev1.ClusterImageSet{
+			ObjectMeta: v1.ObjectMeta{
+				Name: name,
+				Annotations: map[string]string{
+					"version_in": version,
+				},
+			},
+			Spec: hivev1.ClusterImageSetSpec{
+				ReleaseImage: pullspec,
+			},
+		}
+		raw, err := yaml.Marshal(clusterimageset)
+		if err != nil {
+			logrus.WithError(err).Fatalf("Could not marshal yaml for clusterimageset %s", name)
+		}
+		if err := ioutil.WriteFile(filepath.Join(o.outputDir, fmt.Sprintf("%s_clusterimageset.yaml", name)), raw, 0644); err != nil {
+			logrus.WithError(err).Fatalf("Failed to write file for clusterimageset %s", name)
+		}
+	}
+
+	// delete old clusterimagesets
+	for _, path := range toDelete {
+		if err := os.Remove(path); err != nil {
+			logrus.WithError(err).Fatalf("Failed to delete file %s", err)
+		}
+	}
+
+	// update all clusterpool specs
+	for version, files := range autoPools {
+		imagesetName, err := nameFromPullspec(versionToPullspec[version], version)
+		if err != nil {
+			// this shouldn't happen
+			logrus.WithError(err).Fatalf("Failed to generate clusterimageset name for version %s", version)
+		}
+		for _, path := range files {
+			raw, err := ioutil.ReadFile(path)
+			if err != nil {
+				logrus.WithError(err).Fatalf("Failed to read file %s", path)
+			}
+			// unmarshalling a remarshalling the clusterpool object would result in all fields being reordered alpabetically
```

Comment:
Do we just need to determinize them once?

Reply:
Would not want to see us doing regex on yaml

Reply:
I think the determinized clusterpools look messy. Example of the automated pool in the integration tests:

```
apiVersion: hive.openshift.io/v1
kind: ClusterPool
metadata:
  creationTimestamp: null
  labels:
    architecture: amd64
    cloud: aws
    owner: openshift-ci
    product: ocp
    region: us-east-1
    version: "4.5"
    version_in: '>4.5.0-0 <4.6.0-0'
  name: ci-ocp-4-5-amd64-aws-us-east-1
  namespace: ci-cluster-pool
spec:
  baseDomain: hive.aws.ci.openshift.org
  imageSetRef:
    name: ocp-release-4.5.41-x86_64-for-4.5.0-0-to-4.6.0-0
  installConfigSecretTemplateRef:
    name: install-config-aws-us-east-1
  maxSize: 20
  platform:
    aws:
      credentialsSecretRef:
        name: hive-aws-credentials
      region: us-east-1
  pullSecretRef:
    name: pull-secret
  size: 10
  skipMachinePools: true
status:
  ready: 0
  size: 0
```

If you think that's fine, I can change the tool to marshal the clusterpools with automated versioning.

Reply:
Why is this any more or less messy than what we've got in there now?